### ISSUE TITLE
np.cos(X) takes forever (or gets stuck, I'm not sure)

### ISSUE THREAD

**Issue by soerenwolfers** [numpy/numpy#12753]
`np.cos` takes an unreasonable amount of time for an array that differs only slightly from one where it is fast.

### Reproducing code example:

```python
import numpy as np
L = 2**26
N = L*4
x = np.linspace(-L,L,N)
np.cos(np.pi*x)
```
takes 13 seconds.
```python
import numpy as np
L = 2**26
N = L*4+1 # Here is the tiny difference
x = np.linspace(-L,L,N)
np.cos(np.pi*x)
```
takes forever  (I just cancelled it after two hours, it was still running at 100% on one core, memory is not an issue, I have 96GB and it was only using around 4GB).

Even if I use `L = 2**27` in the first example, the runtime only doubles to 28 seconds, as expected.
On the other hand, if I use `L=2**25` in the second example, I do get a result but it takes much longer than `L=2**25` in the first example.

### Numpy/Python version information:

```
>>> import sys,numpy; print(numpy.__version__,sys.version)
1.14.6 3.6.4 |Anaconda, Inc.| (default, Jan 16 2018, 18:10:19) 
[GCC 7.2.0]
```

This is on Ubuntu 18.

**Comment by mattip** [comment]
The NumPy supplied by Anaconda may use Intel-provided patches. Can you see if you have `np.erf` (which is an enhancement added by Intel)? If so, the problem should be reported on the [Anaconda issue tracker](https://github.com/ContinuumIO/anaconda-issues/issues), see [this comment](https://github.com/numpy/numpy/issues/12669#issuecomment-451768026) and a similar issue on Anaconda ContinuumIO/anaconda-issues#10485

**Comment by soerenwolfers** [comment]
@mattip `import numpy; numpy.erf` gives `AttributeError: module 'numpy' has no attribute 'erf'`

**Comment by soerenwolfers** [comment]
I guess this could be related: https://stackoverflow.com/questions/35815093/sine-calculation-orders-of-magnitude-slower-than-cosine

**Comment by seberg** [comment]
If you have the mkl version that seems to apply the Intel patch. But then I would expect `np.erf` or at least `np.core.umath.erf` or so to exist (as well as `np.__mkl_version__` which is likely a better indicator. Maybe you can check if you have an mkl version, otherwise I guess it might be the libc?

**Comment by soerenwolfers** [comment]
@seberg I have none of those

**Comment by oleksandr-pavlyk** [comment]
@soerenwolfers In the interest of reproducing your precise environment, please provide the 'Model name' and 'Flags' from output of `lscpu`, please.

I am not able to reproduce this locally:

```
(issue12753) op@vmlin:~$ ipython
Python 3.6.4 |Anaconda, Inc.| (default, Mar 13 2018, 01:15:57)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.2.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import numpy as np

In [2]: %%time
   ...: import numpy as np
   ...: L = 2**26
   ...: N = L*4+1 # Here is the tiny difference
   ...: x = np.linspace(-L,L,N)
   ...: np.cos(np.pi*x)
   ...:
   ...:
CPU times: user 15.9 s, sys: 10.5 s, total: 26.4 s
Wall time: 1.43 s

In [3]: exit
(issue12753) op@lin:~$ conda list numpy
# packages in environment at ~/miniconda3/envs/issue12753:
#
# Name                    Version                   Build  Channel
numpy                     1.14.6           py36h3b04361_4
numpy-base                1.14.6           py36h81de0dd_4
```

**Comment by oleksandr-pavlyk** [comment]
@soerenwolfers I have used the following command `conda create -n issue12753 -c defaults python=3.6.4 numpy=1.14.6 ipython` to recreate your environment. 

Could you confirm whether you still experience slowness in this environment. If you do not, please post output of ``conda list --explicit``, which would allow me to reproduce your exact environment and hopefully diagnose the issue. If you are not inclined to do so, please feel free to resolve the issue.

**Comment by oleksandr-pavlyk** [comment]
@soerenwolfers Thank you for that information. I was able to recreate your environment and reproduce the hang. 

I used Intel Amplifier VTune to profile the script, where I used `L = 2**22` and `N = 4*L + 1`. Here is the screenshot of hotspots:

![image](https://user-images.githubusercontent.com/21087696/51621045-45e77f00-1ef9-11e9-9690-625747925fb8.png)

The bottleneck is the call to `__cos_avx` (on my machine, which supports AVX vector instructions) to system's runtime math library `libm.so.6`. On your machine it may be a different function specialized to the instructions set of your machine. This function is executed **sequentially**. Your environment uses OpenBLAS linked unpatched NumPy:

```
(slow_cos_issue) [10:25:31 fxsatlin03 slow_cos]$ conda list numpy
# packages in environment at ~/miniconda3_cb3/envs/slow_cos_issue:
#
# Name                    Version                   Build  Channel
numpy                     1.14.6          py36_blas_openblashd3ea46f_200  [blas_openblas]  conda-forge
numpydoc                  0.7.0            py36h18f165f_0
```

The other environment that I was creating trying to reproduce the problem, I used MKL-linked NumPy, which furthermore contains changes to use MKL Vector Math Library to compute transcendental and elementary functions on vectors in **threaded** and vectorized fashion. The execution time drops significantly from 3 seconds to 0.15 seconds on my machine.

The timing of 3 seconds noted in the reconstructed environment is consistent with what I clock in 
 NumPy 1.15 pip-installed into Anaconda environment with CPython and iPython only. 

My investigation indicates that it is not a bug, and definitely does not warrant Intel/Anaconda tag.

**Comment by mattip** [comment]
Thanks @oleksandr-pavlyk for the detailed debug. So the problem is poor performance of cos in `libm.so.6` together with the single-threaded ufunc in vanilla numpy?

**Comment by oleksandr-pavlyk** [comment]
Yes, indeed.

**Comment by soerenwolfers** [comment]
@oleksandr-pavlyk  Even if this is not a bug: Since it acts like one towards an inexperienced user like me, could you summarize for me what I did wrong to get the 'bad' numpy version and how I can switch to the 'good' one? Thanks for your efforts

**Comment by oleksandr-pavlyk** [comment]
@soerenwolfers You did _nothing_ wrong. 

NumPy implements evaluation of a set of functions, like cosine, on ndarrays by calling the corresponding C function, or performing an arithmetic operation, on elements of array in the loop (these loops, implemented in C can be found in [loops.c.src](https://github.com/numpy/numpy/blob/master/numpy/core/src/umath/loops.c.src)).

These loops are all sequential, i.e. they do not use threads, and are executed on a single CPU. 

Intel has authored patches to these loops, which call Intel (R) MKL  [VML functions](https://software.intel.com/en-us/mkl-developer-reference-c-vm-mathematical-functions#0C983A0A-02BA-46A9-AAA5-EAD9D8869ADD) to perform the looping in multi-threaded way, i.e. trying to use all the cores, and make the most of the instruction set of the machine where it is executed. 

This patches are being applied to NumPy shipped in [Intel (R) Distribution for Python*](https://software.intel.com/en-us/distribution-for-python). They are also being applied to MKL-enabled NumPy in [Anaconda Python distribution](https://www.anaconda.com/what-is-anaconda/).

Since NumPy distributed by community does not use MKL library, favoring open-source OpenBLAS. These libraries are used to accelerate basic linear algebra routines and linear algebra solvers, but NumPy itself ships with generic implementations of those, and one can forgo use of any library altogether, which will make things quite a lot slower. Anaconda's NumPy linked against OpenBLAS does not optimized umath loops, hence performances scalar umath functions in OpenBLAS-powered NumPy are mostly similar in no matter the provenance (Anaconda or PyPI). You would get similar performance if you built unpatched MKL-linked NumPy yourself. 

It would befall onto you to investigate the possibility of using MKL-enabled NumPy in your applications. 
To choose MKL enabled NumPy in conda you should do `conda install numpy blas=mkl`. Conda will try solving your package dependency tree and indicate if they can not be satisfied.

**Comment by seberg** [comment]
@soerenwolfers there is nothing wrong with your numpy/python setup. There is probably still a performance bug in your systems `libm`. Oleksandr tested a modified version of numpy which uses MKL for these calls. That may speed things up for your use case (possibly considerably), although I would blame a performance bug in the specific `libm` for the huge difference you see.

**Comment by soerenwolfers** [comment]
@oleksandr-pavlyk  `conda install numpy blas=mkl` gives me "PackagesNotfoundError"   

@seberg so there are users with the same numpy setup as mine but an improved `libm`? How would I change `libm` then without touching numpy? Should I report the problem to the creators of `libm`?

**Comment by charris** [comment]
@soerenwolfers libm gets updated on a regular basis and usually comes with the compiler. On my fedora system
```
charris@fc [numpy.git (master)]$ ll /usr/lib64/libm* | grep '\<libm\>'
-rwxr-xr-x. 1 root root  2080424 Dec 13 17:58 /usr/lib64/libm-2.28.so
-rw-r--r--. 1 root root      141 Dec 13 17:45 /usr/lib64/libm.so
lrwxrwxrwx. 1 root root       12 Dec 13 17:46 /usr/lib64/libm.so.6 -> libm-2.28.so
```
So I have version 2.28, which also matches the version of glibc. It is an historical artifact that libm is separate from glibc,

**Comment by njsmith** [comment]
On Linux systems, libm is part of libc, and on Ubuntu (like almost all Linux systems), libc is maintained by the glibc project and distributed as part of the core OS.

Their bug tracker is here: https://sourceware.org/bugzilla/

This seems like a good bug to file with them. Ideally you'd want a short C program that demonstrates the issue, to make clear that it isn't some weird numpy quirk.

**Comment by soerenwolfers** [comment]
@charris And you don't have the problem? Then an update of `libc` (or of gcc, I guess?) should solve my problem, I have 2.27

**Comment by mattip** [comment]
@soerenwolfers glibc is a core part of the OS, and is rarely updated within the same release. Even Ubuntu 18.10 seems to only upgrade to glibc 2.27.4

**Comment by soerenwolfers** [comment]
@mattip Alright, I will close this issue then. Thanks everyone for your work. By the way, I believe https://sourceware.org/bugzilla/show_bug.cgi?id=13932 might be what solves the issue in future versions of `libc`

**Comment by charris** [comment]
@soerenwolfers Right, I don't see the problem with version 2.28. It might be that it is also fixed in one of the 2.27.x micro version updates, but that would need checking.

**Comment by charris** [comment]
> The GNU C Library releases every 6 months. See the NEWS file in the glibc sources for more information. The current stable version of glibc is 2.28, released on August 1, 2018. The current development version of glibc 2.29, releasing on or around February 1, 2019.
